main.go

```diff
@@ -0,0 +1,176 @@
+			toWPTMetadata(testName[1:], link)
+		}
+	}
+}
+
+// toWPTMetadata converts a pending metadata entry to a wpt-metadata META.yml
+// entry and writes it to disk.
+//
+// Directories and META.yml files are created as needed, or an existing one may
+// be updated.
+func toWPTMetadata(testName string, link shared.MetadataLink) {
+	dirpath := path.Dir(testName)
+	testpath := path.Base(testName)
+	createDirIfNeeded(dirpath)
+
+	// Open the META.yml file that we would write to, if it exists. In the case
+	// where it does exist we don't want to overwrite the file, so open in
+	// read-write-or-create mode.
+	filepath := dirpath + "/META.yml"
+	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if fi.Size() == 0 {
+		// This is an entirely new META.yml file, so create a new
+		// shared.Metadata representation for our expectation and write it out.
+		fmt.Println("CREATED-META-YML: for test", testName, "with url", link.URL)
+		metadata := shared.Metadata{
+			Links: []shared.MetadataLink{
+				{
+					Product: link.Product,
+					URL:     link.URL,
+					Results: []shared.MetadataTestResult{{
+						TestPath: testpath,
+					}},
+				},
+			},
+		}
+		writeToFile(metadata, f)
+		return
+	}
+
+	data, err := ioutil.ReadAll(f)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var metadata shared.Metadata
+	err = yaml.Unmarshal(data, &metadata)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// First check whether we have an existing crbug link in this META.yml; if
+	// so we just want to append our testname to the set of tests associated
+	// with it.
+	hasAdded := false
+	for index, existingLink := range metadata.Links {
+		if !link.Product.MatchesProductSpec(existingLink.Product) {
+			continue
+		}
+
+		// Check whether we are already listed in this entry.
+		for _, result := range existingLink.Results {
+			if result.TestPath == testpath {
+				fmt.Println("EXISTING-LINK: for test", testName, "with url", link.URL)
+				return
+			}
+		}
+
+		if existingLink.URL == link.URL {
+			fmt.Println("APPEND-LINK: for test", testName, "with url", link.URL)
+			metadata.Links[index].Results = append(existingLink.Results, shared.MetadataTestResult{TestPath: testpath})
+			hasAdded = true
+			break
+		}
+	}
+
+	// Otherwise, this is a brand new crbug entry in the META.yml, so add a new
+	// link to it.
+	if !hasAdded {
+		fmt.Println("NEW-LINK: for test", testName, "with url", link.URL)
+		newLink := shared.MetadataLink{
+			Product: link.Product,
+			URL:     link.URL,
+			Results: []shared.MetadataTestResult{{
+				TestPath: testpath,
+			}},
+		}
+		metadata.Links = append(metadata.Links, newLink)
+	}
+
+	writeToFile(metadata, f)
+}
+
+// createDirIfNeeded is a helper function to create a directory if it doesn't
+// already exist. It handles recursive directory creation.
+func createDirIfNeeded(dir string) {
```

Comment:
It looks like just calling [`os.MkdirAll`](https://golang.org/pkg/os/#MkdirAll) would work here, per the documentation it does nothing if the directory already exists.